source/pip/src/interpreter/data_interop.rs

```diff
@@ -0,0 +1,426 @@
+                    let udt_fields = collect_udt_fields(udt).ok()?;
+                    let mut fields = Vec::new();
+
+                    for (name, ty) in udt_fields {
+                        fields.push((name.to_string(), type_ir_from_qsharp_ty(ctx, ty)?));
+                    }
+
+                    Some(TypeIR::Udt(UdtIR {
+                        name: name.to_string(),
+                        fields,
+                    }))
+                }
+            }
+        }
+        Ty::Param { .. } | Ty::Infer(..) | Ty::Arrow(..) | Ty::Err => None,
+    }
+}
+
+pub(crate) fn value_to_pyobj(
+    ctx: &interpret::Interpreter,
+    py: Python,
+    value: &Value,
+) -> PyResult<PyObject> {
+    match value {
+        Value::Int(val) => val.into_py_any(py),
+        Value::BigInt(val) => val.into_py_any(py),
+        Value::Double(val) => val.into_py_any(py),
+        Value::Bool(val) => val.into_py_any(py),
+        Value::String(val) => val.into_py_any(py),
+        Value::Result(val) => {
+            let val = match val {
+                qsc::interpret::Result::Id(_) => {
+                    panic!("unexpected Result::Id in typed_value_to_value_ir")
+                }
+                qsc::interpret::Result::Val(true) => Result::One,
+                qsc::interpret::Result::Val(false) => Result::Zero,
+                qsc::interpret::Result::Loss => Result::Loss,
+            };
+            val.into_py_any(py)
+        }
+        Value::Pauli(val) => {
+            let val = match val {
+                fir::Pauli::I => Pauli::I,
+                fir::Pauli::X => Pauli::X,
+                fir::Pauli::Y => Pauli::Y,
+                fir::Pauli::Z => Pauli::Z,
+            };
+            val.into_py_any(py)
+        }
+        Value::Tuple(values, None) => {
+            let mut tuple = Vec::new();
+            for val in values.iter() {
+                tuple.push(value_to_pyobj(ctx, py, val)?);
+            }
+
+            // Special case Value::UNIT maps to None.
+            if tuple.is_empty() {
+                Ok(py.None())
+            } else {
+                PyTuple::new(py, tuple)?.into_py_any(py)
+            }
+        }
+        Value::Tuple(values, Some(store_item_id)) => {
+            let (udt, kind) = ctx.udt_ty_from_store_item_id(**store_item_id);
+
+            match kind {
+                interpret::UdtKind::Angle => {
+                    let value = values[0].clone().unwrap_int();
+                    let size = values[1].clone().unwrap_int();
+                    let value = u64::try_from(value).expect("value should fit in u64");
+                    let size = u32::try_from(size).expect("size should fit in u32");
+                    let angle = qsc::qasm::stdlib::angle::Angle::new(value, size);
+                    let angle: f64 = angle
+                        .try_into()
+                        .map_err(|_| QasmError::new_err("cannot cast angle to 64-bit float"))?;
```

Comment:
nitpick: "cannot" sounds like it's not allowed, but here I think the error is that we are failing to do it.

```suggestion
                        .map_err(|_| QasmError::new_err("failed to cast angle to 64-bit float"))?;
```

Reply:
Yes, that wording was definitely misleading. Thanks!